coordinator/src/main/scala/filodb.coordinator/QueryActor.scala

```diff
@@ -25,6 +22,88 @@ import filodb.core.store.CorruptVectorException
+// low-level ones, and spawned ones arrive in the Actor mailbox and will go striaght to executing if the parent
+// plan is already executing.
+// A single threaded scheduler is created for each query (all ExecPlans with the same query ID).  This reduces
+// thread contention significantly and reduces task switching overhead, boosting query throughput especially
+// when the tasks are small.
+class QueryScheduler(concurrentQueries: Int) {
+  val currentIDs = new mHashMap[String, Int]().withDefaultValue(0)
+  val queue = new mQueue[(ExecPlan, ActorRef)]
+  val idToSched = new mHashMap[String, Scheduler]()
+  val schedulers = new mQueue[Scheduler]()
+
+  // Given a plan and replyTo, executes the plan if there is room or it is a currently executing query.
+  // If there isn't room, add it to the queue.
+  def execOrWait(plan: ExecPlan,
+                 replyTo: ActorRef,
+                 newPlanFunc: (ExecPlan, ActorRef, Scheduler) => Future[Unit]): Unit = {
+    val id = plan.queryContext.queryId
+    if ((currentIDs contains id) || currentIDs.size < concurrentQueries) {
+      execute(plan, replyTo, newPlanFunc)
+    } else {
+      queue.enqueue((plan, replyTo))
+    }
+  }
+
+  def queueSize: Int = queue.length
+
+  // Pop next item(s?) off top of queue so long as there is room to execute another query.
+  def tryPop(newPlanFunc: (ExecPlan, ActorRef, Scheduler) => Future[Unit]): Unit = synchronized {
+    if (currentIDs.size < concurrentQueries && queue.nonEmpty) {
+      val (nextPlan, replyTo) = queue.dequeue
+      execute(nextPlan, replyTo, newPlanFunc)
+    }
+  }
+
+  // Executes newPlanFunc, updating currentIDs/plans as needed
+  // The closure is called in the current thread, but the callback when Future/plan is completed will not be.
+  // Note that an appropriate single threaded scheduler is pulled off the stack and used.
+  private def execute(newPlan: ExecPlan,
+                      replyTo: ActorRef,
+                      newPlanFunc: (ExecPlan, ActorRef, Scheduler) => Future[Unit]): Unit = {
+    val queryId = newPlan.queryContext.queryId
+
+    // Add plan to data structures and obtain scheduler
+    val sched = synchronized {
+      currentIDs(queryId) += 1
+      idToSched.getOrElseUpdate(queryId, {
+        if (schedulers.nonEmpty) schedulers.dequeue
+        else Scheduler.singleThread(s"Query-Queue-${currentIDs.size}")
+      })
+    }
+
+    val planFut = newPlanFunc(newPlan, replyTo, sched)
+
+    // Remove plan on Future complete.  If a slot opens up, recursively call myself to open up a new slot
+    var newspace = false
+    planFut.onComplete {
+      case _ => synchronized {
+                  currentIDs(queryId) -= 1
+                  if (currentIDs(queryId) == 0) {
+                    currentIDs.remove(queryId)
+                    schedulers.enqueue(idToSched.remove(queryId).get)
+                    newspace = true
+                  }
+                }
+                if (newspace) {
```

Comment:
No need to use newspace var. Can keep the tryPop in the sync block because tryPop is synchronized anyhow.